src/rulesets/Base/fastmath_able.jl

```diff
@@ -36,10 +36,88 @@ let
 
 
         # Unary complex functions
-        @scalar_rule abs(x::Real) sign(x)
-        @scalar_rule abs2(x) 2x
-        @scalar_rule angle(x::Real) Zero()
-        @scalar_rule conj(x::Real) One()
+        ## abs
+        function frule((_, Δx), ::typeof(abs), x::Real)
+            return abs(x), sign(x) * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs), z::Complex)
+            Ω = abs(z)
+            return Ω, (real(z) * real(Δz) + imag(z) * imag(Δz)) / Ω
+        end
+        
+        function rrule(::typeof(abs), x::Real)
+            function abs_pullback(ΔΩ)
+                return (NO_FIELDS, real(ΔΩ)*sign(x))
+            end
+            return abs(x), abs_pullback
+        end
+        function rrule(::typeof(abs), z::Complex)
+            Ω = abs(z)
+            function abs_pullback(ΔΩ)
+                Δu = real(ΔΩ) 
+                return (NO_FIELDS, Δu*z/Ω)
+            end
+            return Ω, abs_pullback
+        end
+
+        ## abs2
+        function frule((_, Δx), ::typeof(abs2), x::Real)
+            return abs2(x), 2x * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs2), z::Complex)
+            return abs2(z), 2 * (real(z) * real(Δz) + imag(z) * imag(Δz))
+        end
+        
+        function rrule(::typeof(abs2), x::Real)
+            function abs2_pullback(Δx)
+                return (NO_FIELDS, 2real(Δx)*x)
+            end
+            return abs2(x), abs2_pullback
+        end
+        function rrule(::typeof(abs2), z::Complex)
+            function abs2_pullback(Δf)
+                Δu = real(Δf)
+                return (NO_FIELDS, 2real(Δu)*z)
+            end
+            return abs2(z), abs2_pullback
+        end
+
+        ## conj
+        function frule((_, Δz), ::typeof(conj), z::Union{Real, Complex})
+            return conj(z), conj(Δz) 
+        end
+        function rrule(::typeof(conj), z::Union{Real, Complex})
+            function conj_pullback(Δf)
+                return (NO_FIELDS, conj(Δf))
+            end
+            return conj(z), conj_pullback
+        end
+
+        ## angle
+        function frule((_, Δz), ::typeof(angle), x::Real)
+            Δx, Δy = reim(Δz)
+            return angle(x), Δy/x  
+        end
+        function frule((_, Δz), ::typeof(angle), z::Complex)
+            x,  y  = reim(z)
+            Δx, Δy = reim(Δz)
+            return angle(z), (-y*Δx + x*Δy)/abs2(z)  
+        end
+        function rrule(::typeof(angle), x::Real)
+            function angle_pullback(Δf)
+                Δu, Δv = reim(Δf)
+                return (NO_FIELDS, im*Δu/x)
+            end
```

Comment:
I would say the correct definition is this:
```suggestion
            angle_pullback(Δf::Real) = (NO_FIELDS, Zero())
            function angle_pullback(Δf::Complex)
                Δu, Δv = reim(Δf)
                return (NO_FIELDS, im*Δu/x)
            end
```
Of course, one may argue what the result for `x = 0` should be, but that's orthogonal to the issue of how to handle `x::Real`, `Δf::Union{Real,Complex}`.

Reply:
If you do the math though, even for real `x` and real `Δf`, the pullback is still complex.

Reply:
> If you do the math though, even for real `x` and real `Δf`, the pullback is still complex.

I don't think this conforms with the convention laid out in https://github.com/JuliaDiff/ChainRulesCore.jl/pull/167. Namely, we treat complex numbers as just another way of representing `ℝ²`. So just as a `ℝ → ℝ` function's pullback shouldn't take an `ℝ²` covector and definitely shouldn't produce an `ℝ²` covector, that should be the same here. Doesn't your argument rely on treating `x::Real` as actually being embedded in the complex numbers? See also https://github.com/JuliaDiff/ChainRules.jl/issues/212.

Another case where this comes up is `sign` (see https://github.com/JuliaDiff/ChainRules.jl/pull/211). `sign` for a complex number normalizes it to the unit circle in the complex plane. Hence, at `z = 1 + 0im`, the cotangent will be parallel to the imaginary axis, i.e. `0 + Δy * im`. If we treat `sign(x::Real)` as embedded in the complex numbers, then our cotangent would look like this as well, when it should be `0`.

I wrote some thoughts on this here: https://github.com/JuliaDiff/ChainRules.jl/issues/212

Reply:
Yeah, I find that convincing, makes sense.